### PR TITLE
Use RegExp.test and String.startsWith when possible

### PR DESCRIPTION
**Pull request by CapacitorSet**
RegExp.prototype.test is marginally more efficient than String.indexOf and String.search, so I used it when possible (i.e. binary comparisons, that would otherwise discard the result of indexOf/search). I did the same with String.startsWith vs `.indexOf === 0`.

### CODE REVIEW

**Comment by gorhill** [comment]
> RegExp.prototype.test is marginally more efficient than String.indexOf and String.search

Source?

File: src/js/1p-filters.js [hunk 1]
```diff
@@ -101,7 +101,7 @@ var handleImportFilePicker = function() {
     if ( file === undefined || file.name === '' ) {
         return;
     }
-    if ( file.type.indexOf('text') !== 0 ) {
+    if ( !file.type.startsWith("test") ) {
```
Comment (CapacitorSet) [inline src/js/1p-filters.js]: Whoops, nice catch. I reviewed the diff a couple times but missed this.

File: platform/chromium/vapi-background.js [hunk 2]
```diff
@@ -940,20 +940,23 @@ vAPI.net.registerListeners = function() {
...
+            if ( mediaExtensions.test(needle) ) {
                 details.type = 'media';
                 return;
             }
 
             // Still need this because often behind-the-scene requests are wrongly
             // categorized as 'other'
-            if ( '.ico.png.gif.jpg.jpeg.webp.'.indexOf(needle) !== -1 ) {
+            var imageExtensions = /\.(?:ico|png|gif|jpg|jpeg|webp)$/;
+            if ( mediaExtensions.test(needle) ) {
```
Comment (adisib) [inline platform/chromium/vapi-background.js]: I think you mean to test imageExtensions here, not mediaExtensions.

**Comment by CapacitorSet** [comment]
The MDN page for [`RegExp.prototype.search`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/search) suggests to use .test when "executing a match simply to find true or false", though it doesn't mention performance; [a jsperf](https://www.jsperf.com/regexp-test-search-vs-indexof/12) shows RegExp.test and RegExp.search to be marginally faster. The results on my machine (FF 50, Ubuntu 16.04):

* `str.indexOf(/(inline|ruby)/) > -1;` - 1Mops/s
* `str.indexOf(precompiledRegex) > -1;` - 2 Mops/s
* `/(inline|ruby)/.test(str);` - 4 Mops/s
* `precompiledRegex.test(str);` - 31 Mops/s

File: src/js/1p-filters.js (hunk 1 above)
Comment (adisib) [inline src/js/1p-filters.js]: It also doesn't match the code convention of using single quotes.

**Comment by adisib** [comment]
String.prototype.indexOf() doesn't take a regular expression as an argument.

**Comment by gorhill** [comment]
> RegExp.prototype.test is marginally more efficient than String.indexOf and String.search

It makes sense that `RegExp.test()` is better than `String.search()`. Not sure why I went with `String.search()` in the instance above. In any case, this is not code executed in a time critical path, that code is executed *only* once when a new tab is created.

The jsperf link provided to substantiate the changes is worthless. The only valid `indexOf` test logically combines two calls to `indexOf`: not something which occurs in the code you changed.

After looking more into it, I find that the blanket statement "more efficient than String.indexOf" does not hold, except for one case. I created my [own benchmark](https://gorhill.github.io/obj-vs-set-vs-map/indexof-vs-regexp.html), with cases closer to the changed code in the current pull request.

Summary:
- Literal haystack / variable needle: RegExp.test is better.
- Variable haystack / literal needle: String.indexOf is better.

This means most of the changes in this pull request won't help, except for the changes in `platform/chromium/vapi-background.js`. However, these changes are not good, you precompiled the RegExp into the function body itself, meaning the compilation occurs every time the function executes -- defeating the most important argument to use precompiled RegExp. Also, if RegExps are to be used for these cases, it's better to rewrite the code to no longer extract needle. I will take care myself of this -- but there is no emergency, again it is an exceptional code path, not one visited for every single network request.

Most of the changes in there are completely outside of time-critical code paths, for example in UI pages, and involve replacing String.indexOf with RegExp.test (which is now debunked). I won't go through time-consuming review of pull requests unless the changes are **measurably** beneficial. In the current case, it's not only non-measurable, the changes accomplish the opposite of the stated goal.

Also, as already noted you did not respect the single-quote convention.

As a result, I will decline this pull request.

Any such pull request will be considered in the future **if and only if** it makes the case that there are _measurable_ benefits -- i.e. _sound_ high-level benchmark results are provided. Examples of such work with _measurable_ benefits:
- <https://github.com/gorhill/uBlock/releases/tag/1.9.8>
- <https://bugzilla.mozilla.org/show_bug.cgi?id=1232354>

**Pull request closed without merging**